bots/tests-scan

```diff
@@ -358,22 +339,70 @@ def cockpit_tasks(api, update, branch_contexts, repo, pull_data, pull_number, sh
-                        priority = github.NO_TESTING
-                        changes = { "description": github.NO_TESTING, "context": repo_context, "state": "pending" }
-                    else:
-                        (priority, changes) = prioritize(status, title, labels, baseline, repo_context, number)
-                    if not update or update_status(api, revision, repo_context, status, changes):
-                        # don't specify a ref to check out; make-source already checks out project master
-                        results.append((priority, "pull-%d" % number, number, revision, "", context, base, proj_repo, "pull/%s/head" % number))
+                    result.append(context + "@" + proj_repo)
+            return result
+
+        def is_valid_context(context):
+            return context.split("@")[0] in contexts
+
+        # Create list of statuses to process
+        todos = {}
+        for status in statuses: # Firstly add all valid contexts that already exist in github
+            if is_valid_context(status):
+                todos[status] = statuses[status]
+        # QUESTION todos or statuses?? if only invalid contexts were triggered then this would trigger all contexts
+        if not todos: # If none defined in github add basic set of contexts
+            for context in branch_contexts.get(base, contexts):
+                todos[context] = {}
+        if trigger_externals():
+            for context in get_externals():
+                if context not in todos:
+                    todos[context] = {}
+
+        for context in todos:
+            # Get correct project and branch. Ones from test name have priority
+            project = repo
+            branch = base
+            (os_scenario, _, override) = context.partition("@")
+            override = override.split("/")
+            if len(override) == 2:
+                project = "/".join(override)
+            elif len(override) == 3:
+                project = "/".join(override[:2])
+                branch = override[2]
+
+            # We need to distinguish between external projects or stable branches,
+            # since those need bots_ref and not ref
+            ref = "pull/%d/head" % number
+            is_stable = base != branch
+            needs_bots = repo != "cockpit-project/cockpit" or is_stable
+
+            # For unmarked and untested status, user must be in whitelist
+            # Not this only applies to this specific commit. A new status
+            # will apply if the user pushes a new commit.
+            status = todos[context]
+            if login not in whitelist and status.get("description", github.NO_TESTING) == github.NO_TESTING:
+                priority = github.NO_TESTING
+                changes = { "description": github.NO_TESTING, "context": context, "state": "pending" }
+            else:
+                (priority, changes) = prioritize(status, title, labels, baseline, context, number)
```

Comment:
`changes` should be `update`

Reply:
https://github.com/cockpit-project/cockpit/blob/master/bots/tests-scan#L371
Can you elaborate a bit more? I admit I did not fully studied how this part works, I just picked it up from ^

Reply:
the condition on line 395 (in your patch, not on master) checks for `update`, which is returned as the second argument in the tuple by prioritize.

```
if login not in whitelist and status.get("description", github.NO_TESTING) == github.NO_TESTING:
    priority = github.NO_TESTING
    update = { "description": github.NO_TESTING, "context": context, "state": "pending" }
else:
    (priority, update) = prioritize(status, title, labels, baseline, context, number)

if not update or update_status(api, revision, context, status, changes):
```

Reply:
In that case there would be error in current (in master) implementation. There is `update` passed as argument to `cockpit_tasks` method. If this is really bug in master, let's fix it there

Reply:
oh sorry, you're right. It's not a bug then, update here is `opts.dry`. So that's fine then.